### PR TITLE
[SofaFramework] Create Sofa.Testing module

### PR DESCRIPTION
**Pull request by fredroy**
~~[DEMO]~~

First step for an unified testing system.

This module gathers:
 - helper/testing files: utilitary to write new tests
 - gtest: API Google test (all tests are based on it)
 - SofaGTestMain (force-load SofaSimulationGraph): template of Main to centralize test-related things.

This module needs to be compiled AFTER SofaSimulation, as SofaGTestMain needs SofaSimulationGraph.

Unfortunately, there is still the mess with the plugin SofaTest, which needs to be solved as well.
______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -190,8 +185,13 @@ add_subdirectory(extlibs)
...
 if(SOFA_BUILD_TESTS OR SOFA_BUILD_RELEASE_PACKAGE)
-    # Depends on SofaSimulation
-    add_subdirectory(tools/SofaGTestMain)
+    add_subdirectory(${SOFA_KERNEL_SOURCE_DIR}/modules/Sofa.Testing)
+    
+    # add tests of previous projects, now that we have Sofa.Testing enabled
+    set(SOFAKERNEL_LIST_TARGET_UNITTESTS SofaHelper SofaDefaultType SofaCore SofaSimulationCore SofaSimulationGraph)
+    foreach(target ${SOFAKERNEL_LIST_TARGET_UNITTESTS})
+        add_subdirectory(${SOFA_KERNEL_SOURCE_DIR}/modules/${target}/${target}_test)
+    endforeach()
```
Comment (guparan) [inline CMakeLists.txt]: remove?

Comment (guparan) [reply CMakeLists.txt]: SofaSimulationGraph_test is missing in tests/CMakeLists.txt though

**Comment by fredroy** [comment]
[ci-build][force-full-build][with-all-tests]

**Comment by fredroy** [comment]
[ci-build][force-full-build][with-all-tests]

File: SofaKernel/modules/Sofa.Testing/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+target_link_libraries(${PROJECT_NAME} PUBLIC SofaHelper SofaDefaultType SofaCore gtest SofaGTestMain SofaSimulationGraph)
+
+target_include_directories(${PROJECT_NAME} PUBLIC "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/${COMPATSRC_ROOT}>")
+target_include_directories(${PROJECT_NAME} PUBLIC "$<INSTALL_INTERFACE:include/Sofa.Testing/Sofa.Testing/compat>")
+source_group("compat" FILES ${COMPAT_HEADER_FILES} )
+
+# own include dir, the macro does not handle "independent" package for SofaFramework-style include 
+# i.e NOT the name of package first (sofa/testing is wanted, instead of Sofa.Testing/)
+target_include_directories(${PROJECT_NAME} PUBLIC "$<BUILD_INTERFACE:${CMAKE_BINARY_DIR}/include/${PROJECT_NAME}/${PROJECT_NAME}>")
+
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Testing/CMakeLists.txt]:
```suggestion
```

File: SofaKernel/modules/Sofa.Testing/SofaGTestMain/CMakeLists.txt [hunk 3]
```diff
@@ -15,3 +15,5 @@ sofa_create_package_with_targets(
     PACKAGE_VERSION ${Sofa_VERSION}
     TARGETS ${PROJECT_NAME} AUTO_SET_TARGET_PROPERTIES
     )
+
+set_target_properties(${PROJECT_NAME} PROPERTIES FOLDER Testing)
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Testing/SofaGTestMain/CMakeLists.txt]:
```suggestion
set_target_properties(${PROJECT_NAME} PROPERTIES FOLDER Testing)

```

File: SofaKernel/modules/Sofa.Type/CMakeLists.txt [hunk 4]
```diff
@@ -57,10 +57,3 @@ sofa_add_targets_to_package(
     INCLUDE_SOURCE_DIR "src"
     INCLUDE_INSTALL_DIR "SofaFramework"
     )
-
-## Tests
-# cmake_dependent_option(SOFA_TYPE_BUILD_TESTS "Compile the automatic tests" ON "SOFA_BUILD_TESTS OR NOT DEFINED SOFA_BUILD_TESTS" OFF)
-# if(SOFA_TYPE_BUILD_TESTS)
-#     enable_testing()
-#     add_subdirectory(${PROJECT_NAME}_test)
-# endif()
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Type/CMakeLists.txt:66]: Let's remove the whole Sofa.Type_test directory then :+1:

File: applications/plugins/SofaPython3/ExternalProjectConfig.cmake.in [hunk 5]
```diff
@@ -2,8 +2,8 @@ cmake_minimum_required(VERSION 3.11)
 
 include(ExternalProject)
 ExternalProject_Add(SofaPython3
-    GIT_REPOSITORY https://github.com/sofa-framework/SofaPython3
-    GIT_TAG origin/master
+    GIT_REPOSITORY https://github.com/fredroy/SofaPython3
+    GIT_TAG origin/follow_sofa_pr1834
```
Comment (guparan) [inline applications/plugins/SofaPython3/ExternalProjectConfig.cmake.in:6]: I see no PR on https://github.com/sofa-framework/SofaPython3 :eyes:

File: applications/projects/Regression/ExternalProjectConfig.cmake.in [hunk 6]
```diff
@@ -2,8 +2,8 @@ cmake_minimum_required(VERSION 2.8.2)
 
 include(ExternalProject)
 ExternalProject_Add(Regression
-    GIT_REPOSITORY https://github.com/sofa-framework/regression
-    GIT_TAG db26140ba6523645627a98c90ed751df4f78cd1c
+    GIT_REPOSITORY https://github.com/fredroy/regression
+    GIT_TAG origin/follow_sofa_pr1834
```
Comment (guparan) [inline applications/projects/Regression/ExternalProjectConfig.cmake.in:6]: You can push your changes directly on https://github.com/sofa-framework/Regression and update GIT_TAG only :+1:

**Review by guparan — CHANGES_REQUESTED** [review]

File: applications/plugins/SofaPython3/ExternalProjectConfig.cmake.in (hunk 5 above)
Comment (fredroy) [reply applications/plugins/SofaPython3/ExternalProjectConfig.cmake.in:6]: I was planning to create PR once everything compiles (and everything is OK) but why not yes

File: applications/projects/Regression/ExternalProjectConfig.cmake.in (hunk 6 above)
Comment (fredroy) [reply applications/projects/Regression/ExternalProjectConfig.cmake.in:6]: Ok i will push directly the branch in the sofa-framework repo from now on then

**Comment by fredroy** [comment]
[ci-build]

**Comment by fredroy** [comment]
[ci-build][force-full-build][with-all-tests]

**Comment by damienmarchal** [comment]
Related to https://github.com/sofa-framework/sofa/issues/1827

**Pull request merged**